cpython-unix/build-cpython.sh

```diff
@@ -858,7 +858,8 @@ def fix_shebang(full):
 
         lines.extend([
             b"#!/bin/sh\n",
-            b'"exec" "\$(dirname \$0)/python${PYTHON_MAJMIN_VERSION}${PYTHON_BINARY_SUFFIX}" "\$0" "\$@"\n',
+            b"'''exec' \"$(dirname -- \"$(realpath -- \"$0\")\")\"/'python${PYTHON_MAJMIN_VERSION}${PYTHON_BINARY_SUFFIX}' \"$0\" \"$@\"\n",
```

Comment:
Oh good point. Mentally I assumed this was an f-string because in the written-out shebang, it seems to use the expanded versions (but clearly it's not). I'll review, modify, and test prior to merging.

Reply:
Ok, so the variables _did_ expand, but the rest of the expressions were also evaluated...

```
#!/bin/sh
'''exec' "."/'python3.9d' "/var/folders/p1/44pzfl0j2m301zzfb0fv0p380000gn/T/tmppqyxz114/build-cpython.sh" ""
' '''
```

Reply:
Wouldn't it make more sense to have this complicated launcher **only if** the Python path contains a space?

Because the Python startup time is already slow enough, I don't think having two extra processes (per startup) (`dirname` and `realpath`) would help much.  (Also it would break if one intends to use this in a minimal container / environment, where the `coreutils` or equivalent aren't present.)

Now I don't exactly know if this launcher should be relocatable, but if so, I think there can be a way to detect in pure POSIX `sh` if the `$0` contains a space, and if so do this, else just leaving it alone.

----

Also, according to the POSIX specification <https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02> one could use something like the following (not tested, but could be worked out):
~~~~
#!/bin/sh
''':' ## this is an alias for `true`
_path="${0}"
_path="${_path%/*}" ## remove the smallest suffix like `/*`
exec -- "${_path}/python3.11d" "${0}" "${@}"
' '''
~~~~

This, should work, and should remove the need for any other additional processes, except the `/bin/sh`.

Reply:
I defer to @paveldikov on the shebang itself.

Reply:
Note however that we _can't_ know if the path will contain a space, since the whole motivation here is that the path is relocatable. The user could put it anywhere!

Reply:
> [...]  we can't know if the path will contain a space [...]

One could use the following simple check (that I believe is POSIX compliant):
~~~~
## replace all spaces and check if the value differs from the original
_path="${0}"
if test "${_path}" == "${_path// }" ; then
    ## the path does not contain a space
else
    ## the path contains a space
fi
~~~~

----

However, the other snippet I've given should work, (and after some extra thought about it) perhaps an extra check should be done to see if the path did contain a `/` or not, as in:
~~~~
#!/bin/sh
''':'
_path_0="${0}"
_path="${_path_0%/*}"
if test "${_path_0}" == "${_path}" ; then
    ## there was no `/` fould in `_path_0`
    _path='.'
fi
exec -- "${_path}/python3.11d" "${0}" "${@}"
' '''
~~~~

Reply:
Happy to track this as a performance improvement